EP039 Blueprint proposal for an event publisher NApp (#557)
kytos-ng/kytos: docs/blueprints/EP039.rst

Motivation
==========

- As Kytos grows and the number of external applications consuming data from Kytos's API increases, maintaining a high-performance system becomes critical. Relying on constant API requests and synchronous data retrieval can overwhelm the application, causing performance degradation. By pushing events to Kafka, external applications can receive the data asynchronously, which reduces the load on Kytos, improves its scalability, and ensures that the system remains fault-tolerant.
- In case of a critical error or debugging cycle, engineers need consistent and reliable logging. By leveraging a lightweight data producer via Kafka, engineers can see exactly what happens in real-time.

Rationale
=========

> The implementation took a simple, object-oriented approach in development. The main reasons behind this were:

Author: Got it, I'll revise this with more emphasis on why certain decisions were chosen, such as why we used Kafka, why the NApp is `async` dominant, etc.

> 

Author: Got it. Once I iron out the details for the blueprint I'll use the data or run some new tests to see average rates and maximum rates.

> 

Author: I pushed the required changes but the benchmark assessment might need some overview. I asked Aldo to help me create a script to simulate a high load by creating 10 threads to constantly send 400 requests to make `kytos/mef_eline.created` events but the results were rather slow.

  1. Encapsulation of Concerns - Each component is self-contained with a clear boundary, reducing unintended side effects and making debugging easier.
  2. Code Reusability & Maintainability - By encapsulating logic within classes and methods, components can be reused when the project is extended or updated without modifying core functionality.
  3. Improved Readability & Organization - Object-oriented code structures the application into well-defined classes, making it easier to navigate, understand, and modify.
  4. Testability - With an object-oriented approach, dependencies can be injected, making unit testing and mocking easier.

I. Requirements
===============

This blueprint has the following characteristics:

>   1. - This NApp requires a reachable Kafka cluster to be fully operational.

Author: Jeronimo was talking with me about this last week and we discussed same names. `kafka_events` was one, but also `publisher`, `exporter`, etc. I think `kafka_events` summarizes the NApp pretty well, so let's choose that.

>   1. - This NApp requires a reachable Kafka cluster to be fully operational.

Reviewer: yes, `kafka_events` seems great, reflects very well what it does and pretty concise.

- Publishing to Kafka
  - Serialization
    - Once filtered, kafka_events will add a task to be completed in the MainThread event loop.
    - Events are serialized using JSON and appended to `aiokafka's` batch to be published to the corresponding Kafka topic. This is to be completed before kafka_events is run.
      - The topic name is configured in `settings.py`.
      - The default is `event_logs`.
      - Kafka ensures these events are durably stored and available for consumption by downstream services.
  - Production
    - Events are sent in batches asynchronously, ensuring maximum efficiency in network requests and processing.
    - We rely on batching to enqueue data in the case of a cluster outage.
    - Each message is sent to the same topic, but divided into their respective partitions based on keys.
      - The amount of partitions is equal to the amount of different event types that need to be consumed.
      - The amount of partitions that can be created safely is relatively high, so short-term to medium-term concerns over scalability are covered.

>         - If scalability becomes a concern, we can split topics are continue creating more partitions.

Reviewer: ```suggestion
        - If scalability becomes a concern, we can split topics and create more partitions.
```

Trivial suggestion. I got confused when I read this phrase. Check it out.

    - Usage: Acts as the central channel for event logging and communication between components.
  - COMPRESSION_TYPE
    - Description: The type of compression applied to messages sent to Kafka.
    - Example: "gzip"
    - Usage: Reduces network load and improves performance by compressing message data.

V. `managers/``
=============

This folder contains the relevant domain specific classes that manage data accordingly.

1. `kafka_ops.py`

- This contains the KafkaSendOperations class, which handles data serialization and transfer to `Kafka`.
- Mainly `async`, following conventions and directives from `kytos` advances in `asyncio`.
  - Uses the `aiokafka` framework instead of `kafka-python` or `kafka-python-ng`.

VI. Events
==========

  1. This section describes the how and why events are to be filtered in kafka_events

  - Filtering
    - Events are to be filtered using regular expressions (regex)
      - This allows for flexibility in rejecting some or all of a given producer's events.
    - Some events may not be necessary to downstream consumers, thus ignoring them would reduce Kafka's burden and improve throughput.

>     - Events can be dynamically added/listed/removed at runtime by interacting with the endpoints `/v1/create`, `/v1/list`, and `/v1/delete` respectively.

Reviewer: Also, how will persistence of these filters configuration look like? This seems to deserve to have a DB collection, be loaded and then CRUD'ed during the runtime, since having to reconfigure after every restart would be a bit inconvenient, complementing part of the ignored events in the settings.py

>     - Events can be dynamically added/listed/removed at runtime by interacting with the endpoints `/v1/create`, `/v1/list`, and `/v1/delete` respectively.

Author: Currently, the immutable Filters (core NApps) are immediately loaded as part of the RULE_SET configuration in settings.py. While I agree manually setting this up would be inconvenient, the NApp automatically loads them upon restart, so the only part that the NApp does not save is anything the user manually inserts from an endpoint. This would include anything that is not a core `Kytos` NApp (`kytos/*` or `kytos.*`), such as from `amlight`.

If this is a problem, then we can circle back and add a DB collection in Mongo. We can store each Filter as a small document (having a `pattern`, `description`, and `mutable` component per document).

>     - Events can be dynamically added/listed/removed at runtime by interacting with the endpoints `/v1/create`, `/v1/list`, and `/v1/delete` respectively.

Author: Got it, I just asked Italo and Jeronimo for their input. I believe if any of the non-core NApps are either always on or very commonly on, then having persistence would likely be the option. Otherwise, if the non-core NApps are transiently added, then having no persistence is likely the option.

:EP: 39
:Title: Exporting Kytos events
:Authors:
    - Austin Uwate <auwate AT fiu DOT edu>
:Created: 2025-02-12

> :Status: Draft

Reviewer: ```suggestion
:Status: Accepted
```
